rust/src/rule_parser/xbits.rs

```diff
@@ -0,0 +1,323 @@
+use std::ffi::{CStr, CString};
+use std::os::raw::c_char;
+
+pub const DETECT_BITS_CMD_SET: u8 = 0;
+pub const DETECT_BITS_CMD_TOGGLE: u8 = 1;
+pub const DETECT_BITS_CMD_UNSET: u8 = 2;
+pub const DETECT_BITS_CMD_ISNOTSET: u8 = 3;
+pub const DETECT_BITS_CMD_ISSET: u8 = 4;
+pub const DETECT_BITS_CMD_NOALERT: u8 = 5;
+pub const DETECT_BITS_CMD_MAX: u8 = 6;
+pub const DETECT_XBITS_TRACK_IPSRC: u8 = 0;
+pub const DETECT_XBITS_TRACK_IPDST: u8 = 1;
+pub const DETECT_XBITS_TRACK_IPPAIR: u8 = 2;
+pub const DETECT_XBITS_TRACK_FLOW: u8 = 3;
+pub const DETECT_XBITS_EXPIRE_DEFAULT: u32 = 30;
+
+pub const VAR_TYPE_IPPAIR_BIT: u8 = 11;
+pub const VAR_TYPE_HOST_BIT: u8 = 8;
+pub const VAR_TYPE_FLOW_BIT: u8 = 5;
+
+pub const BIT_TYPE_XBIT: u8 = 0;
+pub const BIT_TYPE_FLOWBIT: u8 = 1;
+pub const BIT_TYPE_HOSTBIT: u8 = 2;
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct FlowbitStr {
+    fb_str: *const c_char,
+    fb_str_len: u32,
+}
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct DetectBitsData {
+    idx: u32,
+    cmd: u8,
+    tracker: u8,
+    expire: u32,
+    vartype: u8,
+    name: *mut c_char,
+    fb_names: *mut *mut c_char,
```

Comment:
some of the complexity seems to come from a desire to make this data structures accessible to C, meaning we need C types etc. Have you considered making it opague to C and just access it through some utility accessor functions? Then we can have simpler logic in Rust using native rust types.